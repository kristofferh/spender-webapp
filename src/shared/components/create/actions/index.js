import {
  ADD_ITEM_REQUEST,
  ADD_ITEM_SUCCESS,
  ADD_ITEM_FAILURE
} from "../constants";

export const addItemRequest = () => ({
  type: ADD_ITEM_REQUEST
});

export const addItemSuccess = data => ({
  type: ADD_ITEM_SUCCESS,
  data
});

export const addItemFailure = errors => ({
  type: ADD_ITEM_FAILURE,
  errors
});

// Make call to the API.
// @todo: move this to utils.
export const makeRequest = search => {
  return fetch(`http://localhost:3000?query=${search}`).then(response => {
    return response.json();
  });
};

export const fetchResults = search => dispatch => {
  // First dispatch: the app state is updated to inform UI
  // that the API call is starting.
  dispatch(addItemRequest());

  makeRequest(search)
    .then(json => {
      // Second dispatch: return results
      return dispatch(addItemSuccess(json));
    })
    .catch(errors => {
      return dispatch(addItemFailure(errors));
    });
};
