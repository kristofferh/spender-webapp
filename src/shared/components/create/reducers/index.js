import {
  ADD_ITEM_REQUEST,
  ADD_ITEM_SUCCESS,
  ADD_ITEM_FAILURE
} from "../constants";

export default (
  state = {
    isFetching: false,
    item: {},
    errors: []
  },
  action
) => {
  switch (action.type) {
    case REQUEST_RESULTS:
      return Object.assign({}, state, {
        isFetching: true
      });
    case RECEIVE_RESULTS:
      return Object.assign({}, state, {
        isFetching: false,
        tracks: action.tracks
      });
    default:
      return state;
  }
};
