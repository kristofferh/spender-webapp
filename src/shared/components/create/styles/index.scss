.form-group {
  display: block;
  margin-bottom: 15px;

  &--vertical {
    .form-group-label-text {
      display: block;
    }
  }
}
