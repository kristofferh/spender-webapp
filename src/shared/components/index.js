import React from "react";
import { BrowserRouter as Router } from "react-router-dom";

const App = () => (
  <Router>
    <div>Hi</div>
  </Router>
);

export default App;
