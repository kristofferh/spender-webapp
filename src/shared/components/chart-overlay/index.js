import React, { Component } from "react";
import PropTypes from "prop-types";

import { white } from "shared/utils/styles";

import { Overlay, Container, Toggle, Arrow } from "./styles";

export default class ChartOverlay extends Component {
  static propTypes = {
    initialDisplay: PropTypes.bool,
    children: PropTypes.any,
  };

  state = {
    display: this.props.initialDisplay,
    height: "auto",
  };

  constructor(props) {
    super(props);
  }

  componentDidMount() {
    if (this.container) {
      window.addEventListener("resize", this.setDimensions);
      this.setDimensions();
    }
  }

  componentWillUnmount() {
    window.removeEventListener("resize", this.setDimensions);
  }

  setDimensions = () => {
    const offset = this.container.getBoundingClientRect();
    const windowHeight = window.innerHeight;
    this.setState({
      height: windowHeight - offset.top,
    });
  };

  toggle = () => {
    this.setState({
      display: !this.state.display,
    });
  };

  render() {
    return (
      <Overlay ref={(node) => (this.container = node)}>
        <Container height={this.state.display ? this.state.height : 0}>
          {this.props.children}
        </Container>
        <Toggle active={this.state.display} onClick={this.toggle}>
          <Arrow active={this.state.display} color={white} />
        </Toggle>
      </Overlay>
    );
  }
}
