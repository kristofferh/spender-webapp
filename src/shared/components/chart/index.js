import React, { Component } from "react";
import PropTypes from "prop-types";
import { LinearGradient } from "@vx/gradient";
import { AreaClosed, LinePath } from "@vx/shape";
import { curveMonotoneX } from "@vx/curve";
import { GridRows, GridColumns } from "@vx/grid";
import { scaleTime, scaleLinear } from "@vx/scale";
import { extent, max } from "d3-array";

import { Container } from "./styles";

const stock = [
  { date: "2007-04-24T07:00:00.000Z", close: 93.24 },
  { date: "2007-04-25T07:00:00.000Z", close: 95.35 },
  { date: "2007-04-26T07:00:00.000Z", close: 98.84 },
  { date: "2007-04-27T07:00:00.000Z", close: 99.92 },
  { date: "2007-04-30T07:00:00.000Z", close: 99.8 },
  { date: "2007-05-01T07:00:00.000Z", close: 99.47 }
];

// accessors
const xStock = d => new Date(d.date);
const yStock = d => d.close;

class Area extends Component {
  static propTypes = {
    width: PropTypes.number,
    height: PropTypes.number,
    margin: PropTypes.shape({
      top: PropTypes.number,
      bottom: PropTypes.number,
      left: PropTypes.number,
      right: PropTypes.number
    }),
    isResponsive: PropTypes.bool
  };

  static defaultProps = {
    width: 800,
    height: 400,
    margin: {
      top: 0,
      left: 0,
      bottom: 0,
      right: 0
    },
    isResponsive: true
  };

  state = {
    render: !this.props.isResponsive,
    width: this.props.isResponsive ? null : this.props.width
  };

  componentDidMount() {
    if (this.props.isResponsive) {
      this.handleResize();
      window.addEventListener("resize", this.handleResize);
    }
  }

  handleResize = () => {
    if (this.container) {
      const containerSize = this.container.getBoundingClientRect();
      this.setState({ width: containerSize.width, render: true });
    }
  };

  render() {
    const { height, margin } = this.props;

    // bounds
    const xMax = this.state.width - margin.left - margin.right;
    const yMax = height - margin.top - margin.bottom;

    // scales
    const xScale = scaleTime({
      range: [0, xMax],
      domain: extent(stock, xStock)
    });
    const yScale = scaleLinear({
      range: [yMax, 0],
      domain: [0, max(stock, yStock) + yMax / 3]
    });

    return (
      <Container innerRef={n => (this.container = n)}>
        {this.state.render ? (
          <svg
            ref={s => (this.svg = s)}
            width={this.state.width}
            height={height}
          >
            <LinearGradient
              from="#fff"
              to="#fff"
              fromOpacity={0.8}
              toOpacity={1}
              id="stroke"
            />
            <LinearGradient
              from="#fff"
              to="#fff"
              fromOpacity={0.15}
              toOpacity={0}
              fromOffset="-10%"
              toOffset="50%"
              id="fill"
            />
            <GridRows
              lineStyle={{ pointerEvents: "none" }}
              scale={yScale}
              width={xMax}
              strokeDasharray="2,2"
              stroke="rgba(255,255,255,0.3)"
            />
            <GridColumns
              lineStyle={{ pointerEvents: "none" }}
              scale={xScale}
              height={yMax}
              strokeDasharray="2,2"
              stroke="rgba(255,255,255,0.3)"
            />
            <AreaClosed
              data={stock}
              xScale={xScale}
              yScale={yScale}
              x={xStock}
              y={yStock}
              strokeWidth={1}
              stroke={"transparent"}
              fill={"url(#fill)"}
              curve={curveMonotoneX}
            />
            <LinePath
              data={stock}
              xScale={xScale}
              yScale={yScale}
              x={xStock}
              y={yStock}
              stroke={"url(#stroke)"}
              strokeWidth={1}
              curve={curveMonotoneX}
            />
          </svg>
        ) : null}
      </Container>
    );
  }
}

export default Area;
