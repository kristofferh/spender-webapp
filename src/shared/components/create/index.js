import React, { Component } from "react";
import { connect } from "react-redux";

import "./styles";

export class Create extends Component {
  render() {
    return (
      <div>
        <h1>Create</h1>
        <form>
          <label className="form-group form-group--vertical">
            <span className="form-group-label-text">Date</span>
            <input type="date" />
          </label>
          <label className="form-group form-group--vertical">
            <span className="form-group-label-text">Amount</span>
            <input type="number" />
          </label>
          <label className="form-group form-group--vertical">
            <span className="form-group-label-text">Description</span>
            <textarea />
          </label>
          <button type="submit">Submit</button>
        </form>
      </div>
    );
  }
}

export default connect()(Create);
