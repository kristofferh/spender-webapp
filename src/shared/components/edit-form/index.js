import React from "react";
import PropTypes from "prop-types";
import { Field } from "redux-form";
import moment from "moment";
import styled from "react-emotion";

import { required, number } from "shared/utils/validators";

import { Input, Textarea } from "shared/components/form-controls";

const Label = styled.label`
  display: block;
  margin-bottom: 15px;
`;

const EditForm = ({ handleSubmit }) => (
  <form onSubmit={handleSubmit}>
    <Label>
      <Field
        name="date"
        component={Input}
        type="datetime-local"
        format={value => {
          return value
            ? moment(value).format(moment.HTML5_FMT.DATETIME_LOCAL)
            : "";
        }}
        label="Date"
      />
    </Label>
    <Label>
      <Field
        name="amount"
        component={Input}
        type="number"
        attributes={{ step: "any" }}
        label="Amount"
        validate={[required, number]}
      />
    </Label>
    <Label>
      <Field
        name="description"
        component={Textarea}
        label="Description"
        validate={required}
      />
    </Label>
    <button type="submit">Submit</button>
  </form>
);

EditForm.propTypes = {
  handleSubmit: PropTypes.func
};

export default EditForm;
