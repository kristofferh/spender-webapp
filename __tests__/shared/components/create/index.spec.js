import React from "react";

import renderer from "react-test-renderer";

import { Create } from "shared/components/create";

describe("Create snapshot", () => {
  it("should render and match the snapshot", () => {
    const tree = renderer.create(<Create />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
