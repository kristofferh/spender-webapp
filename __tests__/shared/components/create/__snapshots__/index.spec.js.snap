// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Create snapshot should render and match the snapshot 1`] = `
<div>
  <h1>
    Create
  </h1>
  <form
    onChange={[Function]}
    onSubmit={[Function]}
  >
    <label
      className="form-group form-group--vertical"
    >
      <span
        className="form-group-label-text"
      >
        Date
      </span>
      <input
        name="date"
        type="date"
      />
    </label>
    <label
      className="form-group form-group--vertical"
    >
      <span
        className="form-group-label-text"
      >
        Amount
      </span>
      <input
        name="amount"
        step="any"
        type="number"
      />
    </label>
    <label
      className="form-group form-group--vertical"
    >
      <span
        className="form-group-label-text"
      >
        Description
      </span>
      <textarea
        name="description"
      />
    </label>
    <button
      type="submit"
    >
      Submit
    </button>
  </form>
</div>
`;
